### PR TITLE
Adds ability to add http2 configuration option SETTINGS_MAX_HEADER_LIST_SIZE

### PR DESCRIPTION
**Pull request by srikailash**
Signed-off-by: sri kailash <sri.gandebathula@booking.com>

*title*: This PR adds ability to add SETTINGS_MAX_HEADER_LIST_SIZE http2 configuration option

*Description*:
This adds max_header_list_size configuration option that can be communicated to upstream hosts.
please see https://github.com/envoyproxy/envoy/issues/2958.

*Risk Level*: Low

TODO:
1. Need to add tests

### CODE REVIEW

File: api/envoy/api/v2/core/protocol.proto [hunk 1]
```diff
@@ -76,6 +76,12 @@ message Http2ProtocolOptions {
   // window. Currently, this has the same minimum/maximum/default as *initial_stream_window_size*.
   google.protobuf.UInt32Value initial_connection_window_size = 4
       [(validate.rules).uint32 = {gte: 65535, lte: 2147483647}];
+
+  // `SETTINGS_MAX_HEADER_LIST_SIZE <https://tools.ietf.org/html/rfc7540#section-6.5>`_
+  //  This option indicates client the maximum size of header, server is ready to accept
```
Comment (PiotrSikora) [inline api/envoy/api/v2/core/protocol.proto]: This is not at all what this setting is used for, please re-read the RFC.

File: api/envoy/api/v2/core/protocol.proto [hunk 2]
```diff
@@ -76,6 +76,12 @@ message Http2ProtocolOptions {
...
+  //  otherwise it throws an error.
+  google.protobuf.UInt32Value max_header_list_size = 5
+      [(validate.rules).uint32 = {gte: 8000, lte: 16000}];
```
Comment (PiotrSikora) [inline api/envoy/api/v2/core/protocol.proto]: Those brackets don't make any sense. Default is unlimited, yet you're enforcing max of 16000 bytes here?

File: include/envoy/http/codec.h [hunk 3]
```diff
@@ -241,6 +242,8 @@ struct Http2Settings {
   // our default connection-level window also equals to our stream-level
   static const uint32_t DEFAULT_INITIAL_CONNECTION_WINDOW_SIZE = 256 * 1024 * 1024;
   static const uint32_t MAX_INITIAL_CONNECTION_WINDOW_SIZE = (1U << 31) - 1;
+  // defaults to maximum, value in octet
+  static const uint32_t DEFAULT_MAX_HEADER_LIST_SIZE = 8000;
```
Comment (PiotrSikora) [inline include/envoy/http/codec.h]: Default is unlimited, please leave it as such. Also, 8000 bytes default would break half of the Internet.

Comment (srikailash) [reply include/envoy/http/codec.h]: @PiotrSikora, wanted to discuss this. updated.

File: api/envoy/api/v2/core/protocol.proto (hunk 1 above)
Comment (srikailash) [reply api/envoy/api/v2/core/protocol.proto]: @PiotrSikora, updated the comment.

**Comment by srikailash** [comment]
@PiotrSikora, will reopen once i get my other PR merged. :)

**Pull request closed without merging**